### ISSUE TITLE
Troubles compile on arm32v7

### ISSUE THREAD

**Issue by lambadalambda** [akash-akya/vix#138]
This might be the same as #130, but there's no cross compilation here.

When trying to install Vix in an arm32v7 docker container, I get the following error:

```
4.742 ==> vix
4.806 make[1]: Entering directory '/root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/c_src'
6.709 ** (CaseClauseError) no case clause matching: {"arm", "linux", "gnueabihf"}
6.709     /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/build_scripts/precompiler.exs:39: Vix.LibvipsPrecompiled.cast_target/1
6.709     /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/build_scripts/precompiler.exs:30: Vix.LibvipsPrecompiled.url/2
6.709     /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/build_scripts/precompiler.exs:16: Vix.LibvipsPrecompiled.fetch_libvips/0
6.709     (elixir 1.12.3) lib/code.ex:1261: Code.require_file/2
6.720 make[1]: *** [Makefile:84: /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips] Error 1
6.721 make[1]: Leaving directory '/root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/c_src'
6.723 make: *** [Makefile:5: all] Error 2
6.728 could not compile dependency :vix, "mix compile" failed. You can recompile this dependency with "mix deps.compile vix", update it with "mix deps.update vix" or clean it with "mix deps.clean vix"
```

The line `6.709 ** (CaseClauseError) no case clause matching: {"arm", "linux", "gnueabihf"}` is interesting, as the precompilation script expects `{"armv7l", "linux", "gnueabihf"}`.

Setting the correct architecture via `CC_PRECOMPILER_CURRENT_TARGET=armv7l-linux-gnueabihf`, I get the same errors as #130.

```
4.735 ==> vix
4.796 make[1]: Entering directory '/root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/c_src'
6.797
6.797 09:00:43.332 [debug] Fetching https://github.com/akash-akya/sharp-libvips/releases/download/v8.15.0/libvips-8.15.0-linux-armv7.tar.gz
8.987
8.987 09:00:45.528 [debug] Extracting to /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/libvips-8.15.0-linux-armv7.tar.gz
9.895 cc -c -O2 -Wall -Werror -Wextra -Wno-unused-parameter -Wmissing-prototypes -std=c11 -D_POSIX_C_SOURCE=200809L -fPIC -I /usr/local/lib/erlang/erts-12.3.2.13/include -I /usr/local/lib/erlang/usr/include -pthread -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/include -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/lib/glib-2.0/include -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/include/glib-2.0 -o utils.o utils.c
11.10 cc -c -O2 -Wall -Werror -Wextra -Wno-unused-parameter -Wmissing-prototypes -std=c11 -D_POSIX_C_SOURCE=200809L -fPIC -I /usr/local/lib/erlang/erts-12.3.2.13/include -I /usr/local/lib/erlang/usr/include -pthread -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/include -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/lib/glib-2.0/include -I /root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/_build/dev/lib/vix/priv/precompiled_libvips/include/glib-2.0 -o g_object/g_value.o g_object/g_value.c
11.51 g_object/g_value.c: In function ‘set_int64’:
11.51 g_object/g_value.c:95:34: error: passing argument 3 of ‘enif_get_int64’ from incompatible pointer type [-Werror=incompatible-pointer-types]
11.51    95 |   if (!enif_get_int64(env, term, &int64_value)) {
11.51       |                                  ^~~~~~~~~~~~
11.51       |                                  |
11.51       |                                  long int *
11.51 In file included from g_object/../utils.h:4,
11.51                  from g_object/g_value.c:3:
11.51 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif_api_funcs.h:134:88: note: expected ‘ErlNifSInt64 *’ {aka ‘long long int *’} but argument is of type ‘long int *’
11.51   134 | ERL_NIF_API_FUNC_DECL(int,enif_get_int64,(ErlNifEnv*, ERL_NIF_TERM term, ErlNifSInt64* ip));
11.51       |                                                                          ~~~~~~~~~~~~~~^~
11.51 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif.h:345:76: note: in definition of macro ‘ERL_NIF_API_FUNC_DECL’
11.51   345 | #  define ERL_NIF_API_FUNC_DECL(RET_TYPE, NAME, ARGS) extern RET_TYPE NAME ARGS
11.51       |                                                                            ^~~~
11.51 g_object/g_value.c: In function ‘set_uint64’:
11.51 g_object/g_value.c:124:35: error: passing argument 3 of ‘enif_get_uint64’ from incompatible pointer type [-Werror=incompatible-pointer-types]
11.51   124 |   if (!enif_get_uint64(env, term, &uint64_value)) {
11.51       |                                   ^~~~~~~~~~~~~
11.51       |                                   |
11.51       |                                   long unsigned int *
11.51 In file included from g_object/../utils.h:4,
11.51                  from g_object/g_value.c:3:
11.51 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif_api_funcs.h:135:89: note: expected ‘ErlNifUInt64 *’ {aka ‘long long unsigned int *’} but argument is of type ‘long unsigned int *’
11.51   135 | ERL_NIF_API_FUNC_DECL(int,enif_get_uint64,(ErlNifEnv*, ERL_NIF_TERM term, ErlNifUInt64* ip));
11.51       |                                                                           ~~~~~~~~~~~~~~^~
11.51 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif.h:345:76: note: in definition of macro ‘ERL_NIF_API_FUNC_DECL’
11.51   345 | #  define ERL_NIF_API_FUNC_DECL(RET_TYPE, NAME, ARGS) extern RET_TYPE NAME ARGS
11.51       |                                                                            ^~~~
12.74 cc1: all warnings being treated as errors
```

Providing my own libvips and changing the compilation mode to PLATFORM_PROVIDED_LIBVIPS leads to an identical error.

```
4.741 ==> vix
4.804 make[1]: Entering directory '/root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/c_src'
5.005 cc -c -O2 -Wall -Werror -Wextra -Wno-unused-parameter -Wmissing-prototypes -std=c11 -D_POSIX_C_SOURCE=200809L -fPIC -I /usr/local/lib/erlang/erts-12.3.2.13/include -I /usr/local/lib/erlang/usr/include `pkg-config vips --cflags` -o utils.o utils.c
6.168 cc -c -O2 -Wall -Werror -Wextra -Wno-unused-parameter -Wmissing-prototypes -std=c11 -D_POSIX_C_SOURCE=200809L -fPIC -I /usr/local/lib/erlang/erts-12.3.2.13/include -I /usr/local/lib/erlang/usr/include `pkg-config vips --cflags` -o g_object/g_value.o g_object/g_value.c
6.610 g_object/g_value.c: In function ‘set_int64’:
6.612 g_object/g_value.c:95:34: error: passing argument 3 of ‘enif_get_int64’ from incompatible pointer type [-Werror=incompatible-pointer-types]
6.612    95 |   if (!enif_get_int64(env, term, &int64_value)) {
6.612       |                                  ^~~~~~~~~~~~
6.612       |                                  |
6.612       |                                  long int *
6.614 In file included from g_object/../utils.h:4,
6.614                  from g_object/g_value.c:3:
6.614 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif_api_funcs.h:134:88: note: expected ‘ErlNifSInt64 *’ {aka ‘long long int *’} but argument is of type ‘long int *’
6.614   134 | ERL_NIF_API_FUNC_DECL(int,enif_get_int64,(ErlNifEnv*, ERL_NIF_TERM term, ErlNifSInt64* ip));
6.614       |                                                                          ~~~~~~~~~~~~~~^~
6.614 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif.h:345:76: note: in definition of macro ‘ERL_NIF_API_FUNC_DECL’
6.614   345 | #  define ERL_NIF_API_FUNC_DECL(RET_TYPE, NAME, ARGS) extern RET_TYPE NAME ARGS
6.614       |                                                                            ^~~~
6.615 g_object/g_value.c: In function ‘set_uint64’:
6.615 g_object/g_value.c:124:35: error: passing argument 3 of ‘enif_get_uint64’ from incompatible pointer type [-Werror=incompatible-pointer-types]
6.615   124 |   if (!enif_get_uint64(env, term, &uint64_value)) {
6.615       |                                   ^~~~~~~~~~~~~
6.615       |                                   |
6.615       |                                   long unsigned int *
6.615 In file included from g_object/../utils.h:4,
6.615                  from g_object/g_value.c:3:
6.615 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif_api_funcs.h:135:89: note: expected ‘ErlNifUInt64 *’ {aka ‘long long unsigned int *’} but argument is of type ‘long unsigned int *’
6.615   135 | ERL_NIF_API_FUNC_DECL(int,enif_get_uint64,(ErlNifEnv*, ERL_NIF_TERM term, ErlNifUInt64* ip));
6.615       |                                                                           ~~~~~~~~~~~~~~^~
6.615 /usr/local/lib/erlang/erts-12.3.2.13/include/erl_nif.h:345:76: note: in definition of macro ‘ERL_NIF_API_FUNC_DECL’
6.615   345 | #  define ERL_NIF_API_FUNC_DECL(RET_TYPE, NAME, ARGS) extern RET_TYPE NAME ARGS
6.615       |                                                                            ^~~~
7.805 cc1: all warnings being treated as errors
7.812 make[1]: *** [Makefile:75: g_object/g_value.o] Error 1
7.813 make[1]: Leaving directory '/root/.cache/mix/installs/elixir-1.12.3-erts-12.3.2.13/1601325522b57b14bb0cbae710a1c1f4/deps/vix/c_src'
7.816 make: *** [Makefile:5: all] Error 2
```

Here's a Dockerfile to reproduce it:

```Dockerfile
FROM arm32v7/elixir:1.12.3

ENV CC_PRECOMPILER_CURRENT_TARGET=armv7l-linux-gnueabihf
ENV VIX_COMPILATION_MODE=PLATFORM_PROVIDED_LIBVIPS

ADD vipstest.exs /

RUN apt update && apt install -y libvips-dev
RUN mix local.hex --force

RUN elixir /vipstest.exs
```

Here's vipstest.exs

```elixir
Mix.install([{:vix, "~> 0.25.0"}])
alias Vix.Vips.{Image, Operation}

:ok =
  Operation.black!(500, 500, bands: 3)
  |> Image.write_to_file("black.jpg")
```

As the architecture is specified in the docker file, it can be run on any system that support docker using qemu for system architecture emulation.

**Comment by akash-akya** [comment]
This is a bit unusual, so we are getting arch as `arm` from the vm. So you are trying to build an image within qemu emulating arm32v7, correct?

**Comment by lambadalambda** [comment]
yes, that's correct. I ran these on an m2 mac, not sure if that makes a difference. I don't have access to an actual arm32 device, sadly, but I know that other arm64 devices have the same issue when building this arm32 build, for example https://git.pleroma.social/pleroma/pleroma/-/jobs/247111

**Comment by lambadalambda** [comment]
It's `arm-unknown-linux-gnueabihf`

**Comment by akash-akya** [comment]
Thanks. I will take a look as soon as I get time

**Comment by lambadalambda** [comment]
Thank you! If there's anything I can do to help, please let me know.

**Comment by akash-akya** [comment]
@lambadalambda can you please check `dev` branch with `VIX_COMPILATION_MODE=PLATFORM_PROVIDED_LIBVIPS` system env set? 

```elixir
{:vix, github: "akash-akya/vix", branch: "dev"}
```

**Comment by lambadalambda** [comment]
Yes, this fixed it! tested on both libc and musl.

**Comment by akash-akya** [comment]
Cool, I'll merge and cut a release. Thanks!

**Comment by akash-akya** [comment]
Fixed by #139

**Comment by akash-akya** [comment]
Released `v0.26.0`
